File: 113_pathSum2.cpp
// Time Complexity : O(N*N) where N is the number of nodes.
// Space Complexity :O(N*N) where N is the number of nodes of tree
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        if(root == NULL){
            vector<vector<int>> x;
            return x;
        }
        vector<int> path;
        dfs(root, targetSum, path);
        return result;
    }
    void dfs(TreeNode* root, int targetSum, vector<int>& path){
        //base
        if(root == NULL){
            return;
        }
        //logic
        path.push_back(root->val);
        if(root->left == NULL && root->right == NULL){
            if((targetSum - root->val) == 0){
                result.push_back(path);
            }
        }
        dfs(root->left, targetSum-root->val, path);
        dfs(root->right, targetSum-root->val, path);
        //
    }
};

// Time Complexity : O(N) where N is the number of nodes.
// Space Complexity :O(N) where N is the number of nodes of tree
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        if(root == NULL){
            vector<vector<int>> x;
            return x;
        }
        vector<int> path;
        dfs(root, targetSum, path);
        return result;
    }
    void dfs(TreeNode* root, int targetSum, vector<int>& path){//path passed by reference
        //base
        if(root == NULL){
            return;
        }
        //logic
        //action
        path.push_back(root->val);
        if(root->left == NULL && root->right == NULL){
            if((targetSum - root->val) == 0){
                result.push_back(path);
            }
        }
        dfs(root->left, targetSum-root->val, path);
        dfs(root->right, targetSum-root->val, path);
        //backtrack
        path.pop_back();
    }
};

File: 113-pathSum2.java
//Approach-1: Creating multiple new lists everytime
//Overall Time Complexity: O(N^2) Deepsopies for all the nodes
//Overall Space Complexity: O(NH) At a time a node at max can hold h elements
class Solution {
    List<List<Integer>> res;
    
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        res = new ArrayList<List<Integer>>();
        helper(root, new ArrayList<Integer>(), 0, targetSum);
        return res;
    }
    
    public void helper(TreeNode node, List<Integer> path, int sum, int target){
        if(node == null) return;
        
        sum += node.val;
        path.add(node.val);
        if(node.left == null && node.right == null && sum == target){
            res.add(path);
        }
        
        helper(node.left, new ArrayList<>(path), sum, target);
        helper(node.right, new ArrayList<>(path), sum, target);
    }
}


//Approach-2: Using Backtracking
//Overall Time Complexity: O(N) - N = Number of nodes in the tree
//Overall Space Complexity: O(H) - H = is the height of the binary tree
//Here we can use either local or global variable for the path
class Solution {
    List<List<Integer>> res;
    
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        res = new ArrayList<List<Integer>>();
        helper(root, new ArrayList<Integer>(), 0, targetSum);
        return res;
    }
    
    public void helper(TreeNode node, List<Integer> path, int sum, int target){
        if(node == null) return;
        
        sum += node.val;
        path.add(node.val);
        if(node.left == null && node.right == null && sum == target){
            res.add(new ArrayList<>(path));
        }
        
        helper(node.left, path, sum, target);
        helper(node.right, path, sum, target);
        
        path.remove(path.size() - 1);
    }
}